specfile/context_management.py

```diff
@@ -0,0 +1,140 @@
+# Copyright Contributors to the Packit project.
+# SPDX-License-Identifier: MIT
+
+import collections
+import contextlib
+import io
+import os
+import pickle
+import sys
+import tempfile
+import types
+from typing import Any, Callable, Dict, Generator, List, Optional, overload
+
+
+@contextlib.contextmanager
+def capture_stderr() -> Generator[List[bytes], None, None]:
+    """
+    Context manager for capturing output to stderr. A stderr output of anything run
+    in its context will be captured in the target variable of the with statement.
+
+    Yields:
+        List of captured lines.
+    """
+    fileno = sys.__stderr__.fileno()
+    with tempfile.TemporaryFile() as stderr, os.fdopen(os.dup(fileno)) as backup:
+        sys.stderr.flush()
+        os.dup2(stderr.fileno(), fileno)
+        data: List[bytes] = []
+        try:
+            yield data
+        finally:
+            sys.stderr.flush()
+            os.dup2(backup.fileno(), fileno)
+            stderr.flush()
+            stderr.seek(0, io.SEEK_SET)
+            data.extend(stderr.readlines())
+
+
+class GeneratorContextManager(contextlib._GeneratorContextManager):
+    """
+    Extended contextlib._GeneratorContextManager that provides get() method.
+    """
+
+    def __init__(self, function: Callable) -> None:
+        super().__init__(function, tuple(), {})
+
+    def __del__(self) -> None:
+        # make sure the generator is fully consumed, as it is possible
+        # that neither __enter__() nor get() have been called
+        collections.deque(self.gen, maxlen=0)
+
+    def get(self) -> Any:
+        """
+        Fully consumes the underlying generator and returns the yielded value.
+
+        Returns:
+            Value that would normally be the target variable of an associated with statement.
+
+        Raises:
+            StopIteration if the underlying generator is already exhausted.
+        """
+        result = next(self.gen)
+        next(self.gen, None)
```

Comment:
What is the reason for the second `next` call? I am probably missing something

Reply:
The generator yields a value (that's the first `next()`) and then does a clean-up (that's the second). It also raises `StopIteration`, that's why there is the second argument, `None`.

https://github.com/packit/specfile/blob/220170185359ffcd8583dcf7cd546f3c19f48afd/specfile/context_management.py#L131-L138

Reply:
In theory in this case only the first `next()` would be sufficient, as the result is "read-only" anyway, but not exhausting the generator would leave it in suspended state, which would break [expectations](https://github.com/packit/specfile/blob/220170185359ffcd8583dcf7cd546f3c19f48afd/specfile/context_management.py#L110-L115), and also could cause troubles at destruction time (though that should be handled by the [destructor](https://github.com/packit/specfile/blob/220170185359ffcd8583dcf7cd546f3c19f48afd/specfile/context_management.py#L47-L50)).